### ISSUE TITLE
ObjectDisposedException on consumer finalization

### ISSUE THREAD

**Issue by mhowlett** [confluentinc/confluent-kafka-dotnet#577]
librdkafka's `rd_kafka_destroy`, which is called by the .NET Consumer's `Dispose` method will automatically cause the consumer to leave the group if required (a blocking operation). As a side effect of this, the `OnPartitionsRevoked` event will be invoked (and others I think), which will typically contain code that references the `Consumer`. `rd_kafka_destroy` is called via `SafeHandle` via `ReleaseHandle`. The handle is marked as closed prior to the `ReleaseHandle` call, so use of the handle in `OnPartitionsRevoked` is a problem since we have logic that throws `ObjectDisposedException`s conditional on `IsClosed`. I added a `IsDisposed` flag that is set after the `rd_kafka_destroy` call like so:

```
        protected override bool ReleaseHandle()
        {
            Librdkafka.destroy(handle);
            IsDestroyed = true;
            return true;
        }
```

and throw `ObjectDisposedException`s based on this. However, this is causing exceptions like: *A callback was made on a garbage collected delegate of type 'Confluent.Kafka!Confluent.Kafka.Impl.Librdkafka+CommitDelegate::Invoke* since as a side effect, the librdkafka destroy method calls this delegate (in addition to others) and at finalization time, the way the code is written, these will typically not exist.

librdkafka's `rd_kafka_consumer_close`  method behaves similarly to `rd_kafka_destroy`. Calling this immediately prior to disposing the handle (which calls `rd_kafka_destroy`) means that all callbacks are called prior to the `kafkaHandle` being disposed. Then calling `rd_kafka_destroy` doesn't call any callbacks (is this strictly always true?) and all is fine.

But: The above is only working if the user calls `Dispose` explicitly. If an object is finalized (is garbage collected without being disposed), the consumer object is being cleaned up before the librdkafka handles.

In addition to the above, in .NET, the `Dispose` method should strictly never block.

We really need a variant of `rd_kafka_destroy` that is non-blocking, which doesn't trigger any callbacks, and which has the consumer leave the group un-cleanly. This is what should be called by `Dispose`.

I also think we should have a method `Close` on the .net consumer, which exits the group cleanly (and blocks). The user can (optionally) call `Dispose` after this, but it's not required. This is the current operation of the `Dispose` method.

There is a new `IAsyncDisposable` interface in .NET which we could potentially make use of, but this is not great - not compatible with the `using` statement for example, so I don't think we should make use of it.

related to all this, we also want strong semantics around what happens to incomplete long running requests if `Dispose` is called when they are executing (I expect what we have now can be improved upon).

cc: @edenhill, we will probably need a long chat about all this.

**Comment by mhowlett** [comment]
more thoughts: it is possible to address this issue by having the .NET client check to see if the librdkafka handle has been disposed before calling the event handlers (and not call them if it has). this would require some locking over execution of the user defined handler. it also means the `OnPartitionsRevoked` method does not get called on `Dispose` (ideally it should be the way things are). We could also defer marking the handle as destroyed until after the call to `rd_kafka_destroy`, but this is not compatible with using `SafeHandle`, which manages the timing of this for us (and we'd prefer to use this rather than define our own finalizers).

And we still have a blocking `Dispose` method and long running finalization, which we shouldn't.

**Comment by mhowlett** [comment]
@jbouzaglou wrote a related test in #463. Before closing this issue, we should incorporate this into our integration tests and verify it passes.

**Comment by mhowlett** [comment]
There is a PR open on librdkafka that allows the consumer to be destroyed without commit / unsubscribe broker requests. this will allow fast dispose of consumers, which is currently not possible ( https://github.com/edenhill/librdkafka/pull/1932 ).

The finalization issue can be resolved by the client library adjusting GC reference counts to all objects the librdkafka destroy method may require via callbacks (this may be a bit tricky to get right). Alternatively we may remove finalization functionality (this should not be relied on). It's noteworthy that finalization has been depreciated from the Java language.

**Comment by mhowlett** [comment]
Resolved in the 1.0-experimental-9 nuget package.

- referencing a new librdkafka build with non-blocking consumer dispose capability.
- Finalizers are correctly implemented and should never result in events being called `ObjectDisposedException`s.
- `Dispose` methods no longer block.
- We didn't implement `IAsyncDisposable`.
- The `Consumer` has a `Close` method as operating as discussed above.
